### ISSUE TITLE
[New Plugin] Remote-controlled channel-strip

### ISSUE THREAD

**Issue by PROPHESSOR** [lsp-plugins/lsp-plugins#400]
The idea is to create a plugin to turn any DAW to a remote-controlled digital mixer for live setup

## Modules

- Gate
  - Modes: Gate, Ducking
  - Params: Threshold, Range, Attack, Hold, Release, 
- Parametric EQ (8 filters)
  - Filter types: Bell, Hi-Shelf, Hi-Cut, Lo-Shelf, Lo-Cut, Notch
  - Filter controls: On/Off, Gain, Frequency, Quality
- Compressor
  - Controls: Threshold, Ratio, Knee, Attack, Hold, Release, Makeup, Peak/RMS
- Channel control
  - Volume, Solo, Mute, Invert phase, Pan

## Transport protocol

I suggest to use WebSockets as it's supported by the majority of web browsers and provides low-latency data transport. We can also use the OSC protocol as an alternative one, but WebSockets is must have.

Also, a good option is to add support for Mixing Station (an universal mobile digital mixer remote control surface). It operates mostly in OSC protocol but also uses WebSockets for Soundcraft UI series.

**Comment by sadko4u** [comment]
I don't think that implementing a separate channel strip controlled by OSC is a good idea.
Much better solution could be just to add remote control mode to the plugin which can be switched between DAW and OSC. Such solution is much more flexible as it allows to control any plugin with the same instrumentary.
The main problem is, how to avoid clashing with DAW which is responsible for controlling input parameters of the plugin. As a simpliest solution, we can force the plugin to ignore the changes of input parameters if OSC mode is enabled. And the value of parameters then can be synchronized by the UI.

**Comment by sadko4u** [comment]
Once again, my suggestions about this are the following: it is much better to introduce the plugin that will act as a OSC server and receive requests.
Other plugins can work as clients to this server using, for example, shared memory buffers. This will turn the OSC server into OSC router: when it receives an OSC event, it converts it, looks for the matching client and routes to the corresponding plugin.  The overall schema should look like this:

![изображение](https://github.com/lsp-plugins/lsp-plugins/assets/12834464/8e412bc9-ff99-490b-b4a3-234a81978444)

**Comment by PROPHESSOR** [comment]
> I don't think that implementing a separate channel strip controlled by OSC is a good idea. Much better solution could be just to add remote control mode to the plugin which can be switched between DAW and OSC.

Adding the ability to manage plugins via OSC has nothing to do with what I suggested
You mentioned that plugins are meant to be controlled by a DAW and this is usually done that way.
For example, REAPER allows you to control plug-in parameters via MIDI and OSC protocols
This is convenient when you need to assign some controller knob to some plugin
But this will work for a specific project setup (the right plugin must be in the right place for REAPER to find it and map it correctly

My idea was completely different
The idea is to create a simulated digital mixing console with remote control that would look the same in any DAW and in any project, differing only in the number of channels
You connect to it from your phone, computer, iPad or anywhere else and you can control all the channels of the project on which the Slave plug-in is installed, as if connected to a digital mixer like Behringer X-Air or Soundcraft UI

And that’s why this architecture is chosen in my proposal.
There must be a channel strip on the channels, and there must be a master node to which clients will connect

**Comment by PROPHESSOR** [comment]
> Once again, my suggestions about this are the following: it is much better to introduce the plugin that will act as a OSC server and receive requests.
Other plugins can work as clients to this server using, for example, shared memory buffers. This will turn the OSC server into OSC router: when it receives an OSC event, it converts it, looks for the matching client and routes to the corresponding plugin. The overall schema should look like this:

The proposal to use shared memory for fast communication between plugins sounds good
But in this case there is no need to implement remote control in all LSP plugins
- Firstly, it's a lot of work and a lot of pitfalls, since they were made to be controlled through a DAW 
- Secondly, for each of the plugins you will need to create a remote control UI and support it with each update
- Well, thirdly, it’s simply not needed by anyone (there will be cases, but they will require complex configuration for use, then it’s better to use the built-in capabilities of the DAW for remote control)

Here we need to make a new plugin that will have a stable API for management and the version of which will make the UI

**Comment by PROPHESSOR** [comment]
![image](https://github.com/lsp-plugins/lsp-plugins/assets/19251320/07f7d823-2fca-4120-add3-8e817c2b8fae)

**Comment by sadko4u** [comment]
The idea of console is different of what LSP Plugins are trying to reach.
You're suggesting to introduce yet another DAW with functions reduced to the mixing console.
Instead of this, you can get the similar result just by using Carla and proper OSC router.
The implementation of such solution requires a lot of man-hours but the final result is doubtful.
Actually, you ask to re-implement the whole set of available plugins so they could be integrated into one another single plugin as it Carla does when it acts as a plugin.

In my schema, the plugin does not need to know anything about the OSC control and we don't need to adopt it to support OSC. Instead of this, the plugin wrapper is responsible to handling OSC events and setting up the parameters of the plugin. You can build any configuration in your DAW of any vendor and make remote-controllable only those plugins that you want to make remote-controllable. Just connect them to the OSC Dispatcher, assign unique identifier and viola, it works.

**Comment by PROPHESSOR** [comment]
Your option is exactly the one that can be realized with Carla. I, in turn, propose to make an all-in-a-box solution for realizing a convenient virtual digital mixer for use during live performances. It can be distributed separately from the LSP plugins bundle (e.g. as a separate product on the website) and even be paid for if necessary. But it is a solution that will greatly simplify live performance work and may even revolutionize the field.

The main point of the project is not to add remote control support to existing plugins, because this can be achieved with built-in tools of the same REAPER (and ReaLearn) and Carla. The point of the project is to provide a standardized, easy to use and convenient solution for implementing a virtual mixer for live performances that has both client and server. This will help to save a lot of money on buying an external mixing console, which in turn has more limited capabilities (especially the number of channels).

**Comment by sadko4u** [comment]
According to my knowledge, changing legacy LV2 ports by plugin is not possible in LV2 at this time.
I've created a pull request in hope that it will be reviewed and merged:

https://github.com/lv2/lv2/pull/72
